Yoast/wordpress-seo
# Display the "blog not public" notification where relevant

## Context


* The site (a.k.a. "blog") not public notification is so critical that we want it to be a site wide admin notice.

## Summary


This PR can be summarized in the following changelog entry:

* Makes the "You're blocking access to robots" notification site-wide.

## Relevant technical choices:

- Uses AJAX to permanently dismiss the notice via a `ignore_search_engines_discouraged_notice` option
- restores the option validation 
- deprecates `blog_public_notice()`
- adds an upgrade routine to remove the previous notification from the Notification Center
- **Important: the version used for deprecation and upgrade routine is set to 14.1*

## Decisions to be made:
**Update: decisions were made. Please skip this part and jump tp Test Instructions.**
~Before I add test instructions and before CR / acceptance, there are a couple things I need feedback from product / architects related to the notice displayed in Gutenberg:~

**1: For product: Consider to not use HTML for the Gutenberg notice**
Our notice copy is supposed to contain some strong text and a link to the Reading Settings page. A custom component for the notice text would be cool but the Gutenberg notices don't allow components to be passed for their content, which is _meh_ 😞 The message passed to the `createNotice` action is casted to a string: https://github.com/WordPress/gutenberg/blob/c5fab6d23ff7478096942412bc20dc30654ca82e/packages/notices/src/store/actions.js#L57-L60

By using `__unstableHTML` the string can contain HTML, with two downsides:
- the layout is broken, see https://github.com/WordPress/gutenberg/issues/21316
- more importantly: unstable props are subject to removal without notice, which exposes us to a breakage at any moment

For these reasons I'd suggest to use plain text and remove both the strong and the link.

Notice with HTML in WordPress 5.3:

<img width="1047" alt="notice with html" src="https://user-images.githubusercontent.com/1682452/78124060-6a00f300-740f-11ea-82b9-64ea8b5cc6c7.png">

Notice with HTML in WordPress 5.4 with the Gutenberg plugin activated:

<img width="1036" alt="notic with html Gutenplugin" src="https://user-images.githubusercontent.com/1682452/78124114-81d87700-740f-11ea-84ab-d1fb7270f09d.png">

Notice with no HTML:

<img width="1052" alt="notice no html" src="https://user-images.githubusercontent.com/1682452/78124136-8a30b200-740f-11ea-8f97-693b0f72bdfd.png">

**1: For Software architects: Where the new code should be placed**
For now, I put the new code within `js/src/edit.js`. This doesn't seem ideal to me for a couple reasons:
- it would be better to import the two new functions from another file and just call it in the Edit class
- considering that in the future we may need to add more Gutenberg Notices, I guess it would be better to have them grouped in a dedicated directory / file.

However, I couldn't think of a proper place. This isn't exactly a "component". It uses `wp.data.dispatch( "core/notices" )` and related `createNotice` action. Any idea where to put such a thing?



## Test instructions

This PR can be tested by following these steps:

- go to WordPress > Settings > Reading Settings
- check the "Discourage search engines from indexing this site" checkbox
- verify the admin notice is printed out only on these pages:
  - WordPress Dashboard
  - Plugins page
  - Update core page
  - Yoast SEO settings pages
- verify the link ''go to your Reading Settings" works
- click the link "I don't want this site to show in the search results." and verify an AJAX request is triggered
- check the admin notice doesn't appear any longer on all the above pages
- optional: check in the database `wp_options` table > `wpseo` option, there's a value `"ignore_search_engines_discouraged_notice";b:1;`

Note: if you need to make the admin notice appear again for testing purposes, put this code somewhere in a place where it is executed: `WPSEO_Options::set( 'ignore_search_engines_discouraged_notice', false );`
- don't forget to remove it 😛 

## UI changes

## Documentation

## Quality assurance


Fixes https://github.com/Yoast/wordpress-seo/issues/14566
